odoo/addons/base/models/res_partner.py

```diff
@@ -354,7 +354,7 @@ def _get_complete_name(self):
         if self.company_name or self.parent_id:
             if not name and self.type in displayed_types:
                 name = type_description[self.type]
-            if not self.is_company:
+            if not self.is_company and not self.env.context.get('partner_display_name_hide_company'):
```

Comment:
Note: I would have moved directly into _compute_display_name (so that complete name stays as it is) (aka: before setting name as complete name in the compute). Not critical, can be done as followup if you agree.